### ISSUE TITLE
Cannot use Kerberos Authentication over a tunnel from Linux as SPN is being generated wrong

### ISSUE THREAD

**Issue by twiggy** [microsoft/azuredatastudio#25853]
- Azure Data Studio Version: 1.49.1
- OS Version: Linux krb5-user/jammy-updates,jammy-security,now 1.19.2-2ubuntu0.4 amd64 [installed]

Steps to Reproduce:

1. use an ssh tunnel to connect to sql server such as ```ssh -L 1433:SQL01.NOTREAL.NET:1433 user@mybastion ```
1a. Note that SQL01.NOTREAL.NET might just be an IP address since DNS resolution may not be possible.
2. Connect to SQL Server from Azure data studio using window authentication.
3. setup krb5.conf and test via kinit and use knvo to verify that your SQLs SPN is all good to go. This is pretty involved so will leave out, but assume SQL SPN is good.
3. Add an entry to you hosts file so that SQL01.NOTREAL.NET
4. Use KRB5_TRACE to get detailed output. ex. ```KRB5_TRACE=./krb5_trace.log azuredatastudio --log trace --verbose```
5. Connect with a connection string such as ```Server=tcp:SQL01.NOTREAL.NET:1433;Trusted_Connection=True;TrustServerCertificate=True;ServerSpn=MSSQLSvc/sql01.NOTREAL.NET@NOTREAL.NET```

It appears that in krb5_trace.log requests are made like ```[631642] 1723839764.552605: Requesting tickets for MSSQLSvc/localhost:1433@NOTREAL.NET, referrals on```

It seems like something is reversing the SQL01.NOTREAL.NET into localhost. 

I've tried rdns=false in my krb5.conf, but it doesn't seem to do anything. It's possible that KRB5 is the one doing this. Since kvno works with ```kvno MSSQLSvc/sql01.notreal.net:1433``` it makes me think its not krb5...

This is all complicated when adding ports. I would think the easiest thing would be for Azure Data Studio to honor ServerSPN= on the connection string vs trying to do a reverse DNS lookup and not bother building the SPN which could be off when not using 1433.

Of course I could be missing something on the connection string to make it all work. I've seen many examples for JDBC driver using SeverSPN=, but not a lot for .NET, etc.

There error you get from Azure is below and kind of leads you down the SQL missing SPN route, thankful for KRB5_TRACE

```Microsoft.Data.SqlClient.SqlException (0x80131904): Cannot authenticate using Kerberos. Ensure Kerberos has been initialized on the client with 'kinit' and a Service Principal Name has been registered for the SQL Server to allow Kerberos authentication.
ErrorCode=InternalError, Exception=Interop+NetSecurityNative+GssApiException: GSSAPI operation failed with error - Unspecified GSS failure.  Minor code may provide more information (Server not found in Kerberos database).
```

Thank you for time and effort.

Does this issue occur when all extensions are disabled?: Yes

**Comment by twiggy** [comment]
Full stack might be useful. ServerSPN on the connection string was added to SqlClient. But when tracing through Microsoft.Data.SqlClient.SNI.SNIProxy.GenSspiClientContext starting to think it didn't get everywhere it needed to be.

https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs

```
 try
            {
                hostEntry = Dns.GetHostEntry(hostNameOrAddress);
            }
            catch (SocketException)
            {
                // A SocketException can occur while resolving the hostname.
                // We will fallback on using hostname from the connection string in the finally block
            }
            finally
            {
                // If the DNS lookup failed, then resort to using the user provided hostname to construct the SPN.
                fullyQualifiedDomainName = hostEntry?.HostName ?? hostNameOrAddress;
            }

            string serverSpn = SqlServerSpnHeader + "/" + fullyQualifiedDomainName;
```
seems like I'm falling into this finally block that is just building with SqlServerSpnHeader + "/" ....

so this might be an issue that I need to put in for SqlClient and not azure studio,but obviously very related.

```
Microsoft.Data.SqlClient.SqlException (0x80131904): Cannot authenticate using Kerberos. Ensure Kerberos has been initialized on the client with 'kinit' and a Service Principal Name has been registered for the SQL Server to allow Kerberos authentication.
ErrorCode=InternalError, Exception=Interop+NetSecurityNative+GssApiException: GSSAPI operation failed with error - Unspecified GSS failure.  Minor code may provide more information (Server not found in Kerberos database).
   at System.Net.Security.NegotiateStreamPal.GssInitSecurityContext(SafeGssContextHandle& context, SafeGssCredHandle credential, Boolean isNtlm, SafeGssNameHandle targetName, GssFlags inFlags, Byte[] buffer, Byte[]& outputBuffer, UInt32& outFlags, Int32& isNtlmUsed)
   at System.Net.Security.NegotiateStreamPal.EstablishSecurityContext(SafeFreeNegoCredentials credential, SafeDeleteContext& context, String targetName, ContextFlagsPal inFlags, SecurityBuffer inputBuffer, SecurityBuffer outputBuffer, ContextFlagsPal& outFlags)
   at Microsoft.Data.SqlClient.SNI.SNIProxy.GenSspiClientContext(SspiClientContextStatus sspiClientContextStatus, Byte[] receivedBuff, Byte[]& sendBuff, Byte[][] serverName)
   at Microsoft.Data.SqlClient.SNI.TdsParserStateObjectManaged.GenerateSspiClientContext(Byte[] receivedBuff, UInt32 receivedLength, Byte[]& sendBuff, UInt32& sendLength, Byte[][] _sniSpnBuffer)
   at Microsoft.Data.SqlClient.TdsParser.SNISSPIData(Byte[] receivedBuff, UInt32 receivedLength, Byte[]& sendBuff, UInt32& sendLength)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.SSPIError(String error, String procedure)
   at Microsoft.Data.SqlClient.TdsParser.SNISSPIData(Byte[] receivedBuff, UInt32 receivedLength, Byte[]& sendBuff, UInt32& sendLength)
   at Microsoft.Data.SqlClient.TdsParser.SSPIData(Byte[] receivedBuff, UInt32 receivedLength, Byte[]& sendBuff, UInt32& sendLength)
   at Microsoft.Data.SqlClient.TdsParser.TdsLogin(SqlLogin rec, FeatureExtension requestedFeatures, SessionData recoverySessionData, FederatedAuthenticationFeatureExtensionData fedAuthFeatureExtensionData, SqlConnectionEncryptOption encrypt)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.Login(ServerInfo server, TimeoutTimer timeout, String newPassword, SecureString newSecurePassword, SqlConnectionEncryptOption encrypt)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.AttemptOneLogin(ServerInfo serverInfo, String newPassword, SecureString newSecurePassword, Boolean ignoreSniOpenTimeout, TimeoutTimer timeout, Boolean withFailover)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.LoginNoFailover(ServerInfo serverInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString connectionOptions, SqlCredential credential, TimeoutTimer timeout)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.OpenLoginEnlist(TimeoutTimer timeout, SqlConnectionString connectionOptions, SqlCredential credential, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance)
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity, SqlConnectionString connectionOptions, SqlCredential credential, Object providerInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString userConnectionOptions, SessionData reconnectSessionData, Boolean applyTransientFaultHandling, String accessToken, DbConnectionPool pool)
   at Microsoft.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection, DbConnectionOptions userOptions)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(DbConnectionPool pool, DbConnection owningObject, DbConnectionOptions options, DbConnectionPoolKey poolKey, DbConnectionOptions userOptions)
   at Microsoft.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.WaitForPendingOpen()
--- End of stack trace from previous location ---
   at Microsoft.Data.SqlClient.SqlRetryLogicProvider.ExecuteAsync(Object sender, Func`1 function, CancellationToken cancellationToken)
   at Microsoft.Data.SqlClient.SqlRetryLogicProvider.ExecuteAsync(Object sender, Func`1 function, CancellationToken cancellationToken)
   at Microsoft.SqlTools.ServiceLayer.Connection.ReliableConnection.ReliableSqlConnection.<>c__DisplayClass31_0.<<OpenAsync>b__0>d.MoveNext() in /_/src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableSqlConnection.cs:line 342
--- End of stack trace from previous location ---
   at Microsoft.SqlTools.ServiceLayer.Connection.ConnectionService.TryOpenConnection(ConnectionInfo connectionInfo, ConnectParams connectionParams) in /_/src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs:line 712
ClientConnectionId:fe4a7bc8-15ef-47f6-a474-26f3217b2aef
```

**Comment by twiggy** [comment]
![image](https://github.com/user-attachments/assets/98e3bd53-17dc-4165-9450-4f7be2f43dfb)

I'm still stepping through this, but it seems like even though a connection string is being used in the GUI its being parsed and then being sent along through to connection. So it would seem Azure Studio would have to be aware of all connection string properties.

If I put in something like Ser1verSPN with a typo I get an error, so I think the connection string is being parsed by SqlTools and SqlClient properly, but then ServerSPN is just never getting through to the driver so I can't override it to work around DNS/SPN/Kerberos behaviors.

![image](https://github.com/user-attachments/assets/ec43c64e-1a2f-453d-af3a-e1503dd4fda0)

Digging into the buildConnectionInfo RPC call I see it effectively loosing the ServerSPN. I'm not sure if the connectionString is being used at all after hitting connect, which is a bit confusing as an option when setting up new connections as it does follow the general way connection strings like jdbc string or odbc strings are used on servers and other applications. Testing a simpler setup I see that a connection string is just parsed and the saved connection fills out all the properties.

So in short I guess I need ServerSPN and FailoverPartnerSPN as described in https://learn.microsoft.com/en-us/sql/connect/oledb/features/service-principal-name-spn-support-in-client-connections?view=sql-server-ver16 added.

**Comment by twiggy** [comment]
Also found a "workaround" to this issue. My local machine has a 192.x address .

So in etc/hosts put my lan ip like ```192.1.2.3 fqdn.ad.corp```

I did some test .NET code and found that ```hostEntry = Dns.GetHostEntry(hostNameOrAddress);`` is probably getting an ip address in SqlClient:SNIProxy.cs vs a fqdn ... so gotta trick it into returning the fqdn and not localhost.

From their the SPN gets built properly and goes through.
